daphne/src/testing.rs

```diff
@@ -470,32 +474,156 @@ impl DapLeader<BearerToken> for MockAggregator {
+        let collect_uri = task_config
+            .leader_url
+            .join(&format!(
+                "/collect/task/{}/req/{}",
+                collect_req.task_id.to_base64url(),
+                collect_id.to_base64url(),
+            ))
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+
+        Ok(collect_uri)
     }
 
+    // Called to retrieve completed CollectResp at the request of Collector.
     async fn poll_collect_job(
         &self,
-        _task_id: &Id,
-        _collect_id: &Id,
+        task_id: &Id,
+        collect_id: &Id,
     ) -> Result<DapCollectJob, DapError> {
-        unreachable!("not implemented");
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        let leader_state = leader_state_store
+            .get(task_id)
+            .ok_or_else(|| DapError::fatal("cannot get collect job for task_id"))?;
+        if leader_state.collect_jobs.contains_key(collect_id) {
+            if let Some(collect_resp) = &leader_state
+                .collect_jobs
+                .get(collect_id)
+                .unwrap()
+                .collect_resp
+            {
+                Ok(DapCollectJob::Done(collect_resp.clone()))
+            } else if leader_state
+                .collect_jobs
+                .get(collect_id)
+                .unwrap()
+                .collect_resp
+                .is_none()
+            {
+                Ok(DapCollectJob::Pending)
+            } else {
+                Ok(DapCollectJob::Unknown)
+            }
+        } else {
+            Err(DapError::fatal("cannot get collect job for collect_id"))
+        }
     }
 
+    // Called to retrieve pending CollectReq.
     async fn get_pending_collect_jobs(
         &self,
-        _task_id: &Id,
+        task_id: &Id,
     ) -> Result<Vec<(Id, CollectReq)>, DapError> {
-        unreachable!("not implemented");
+        let mut leader_state_store_mutex_guard = self
+            .leader_state_store
+            .lock()
+            .map_err(|e| DapError::Fatal(e.to_string()))?;
+        let leader_state_store = leader_state_store_mutex_guard.deref_mut();
+
+        let leader_state = leader_state_store
+            .get_mut(task_id)
+            .ok_or_else(|| DapError::fatal("cannot get collect job for task_id"))?;
+        let mut res: Vec<(Id, CollectReq)> = Vec::default();
+        while !leader_state.collect_ids.is_empty() {
+            let collect_id = leader_state.collect_ids.pop_front().unwrap();
+            res.push((
+                collect_id.clone(),
+                leader_state
+                    .collect_jobs
+                    .get(&collect_id)
+                    .unwrap()
+                    .collect_req
+                    .as_ref()
+                    .unwrap()
+                    .clone(),
```

Comment:
This long list of calls is what we call "code smell": It's highly likely that this code can be simplified in a way that makes it more readable.

Reply:
Agreed. The code looks much cleaner now.